utils/spark_utils.py

```diff
@@ -759,3 +740,64 @@ def get_standard_concept_ids(spark, concept_ids):
             WHERE ca.concept_id_1 IN ({concept_ids})
         """.format(concept_ids=','.join([str(c) for c in concept_ids])))
     return standard_concept_ids
+
+
+def link_events_with_visits(patient_event):
+    """
+    Links events without a visit_occurrence_id to a same day visit
+    """
+    window = W.partitionBy(['cohort_member_id', 'person_id', 'date'])
+    patient_event = patient_event.withColumn('visit_occurrence_id',
+                                             F.last('visit_occurrence_id', ignorenulls=True).over(window)) \
+        .withColumn('visit_concept_id', F.last('visit_concept_id', ignorenulls=True).over(window))
+    return patient_event
+
+
+def create_visits(patient_event, link_events_visits, gap):
+    """
+    Creates visits as a consecutive sequence of events.
+    The parameter gap controls what events are considered consecutive. A gap of 1
+    means events with 0 days between them are consecutive. A gap of 0 means only same day events
+    are consecutive.
+    """
+    # create date_diff column
+    window = W.partitionBy("cohort_member_id", "person_id").orderBy("date")
+    patient_event = patient_event.withColumn("date_difference",
+                                             F.datediff(patient_event.date,
+                                                        F.lag(patient_event.date, 1).over(
+                                                            window)))
+    # first row per person_id gives nan for date_difference, replace with value 2
+    patient_event = patient_event.na.fill(2, "date_difference")
+
+    if link_events_visits:
```

Comment:
I think `if` block (when `link_events_visits is True`) already covers the cases captured by `else`, my understanding is that we only want to create `visits` when `visit_occurrence_id` is missing.  However, it looks like the `else` block creates `visits` and ignores the existing `visit_occurrence_id`, is that the behavior you expected?

Reply:
Yes. In case of very few events with visit info I wanted to have the option to overwrite the visit info with the created one. So the grouping into "visits" is consistent. But for combination of linking and creating visits the existing info would be preserved. 

But maybe it would be better to add some kind of overwrite option. So you would have:

- Link visits
- Link visits and create visits from orphans (preserving visit info that is there)
- Create visits from orphans and preserve existing visit info.
- Create visits for all events and overwrite existing visit info

Reply:
We could automatically apply the following logic for each event:
1. If visit_occurrence_id is present, use that to link the event to the visit.
2. If visit_occurrence_id is not present, but there is a visit on the same date, link to that.
3. If visit_occurrence_id is not present, and there is no visit on the same date, create a visit.

That should work for most databases, right?

Reply:
@schuemie that is what you get with this PR with by using both options `--link_events_to_visits` and `--create_visits_from_dates`. 

Additionally I wanted to be able to create visits for every event to impose my own structure on the sequences. By using `gap` of 1 I would get sequences of events on consecutive days and still have the time aware tokens to represent the gaps between those sequences. 

Maybe it's cleanest if I use a separate option for that?

Reply:
@egillax  got it. To check my understanding, when the overwrite is enabled, we essentially ignore all the visits and impose our own structure, and create some `visits`. Sometimes we might want to use a larger gap to define what a visit means since it's just a way to group events. I can imagine there are cases, where we want to link the events that occurred within 10 days of each other, which would create some `meta` visits. Is my understanding correct?

Reply:
@ChaoPang yes you are correct. You could also look at it as a way to control for how small gaps you'll insert the time aware tokens.